File: zmin9/18회/하_괄호.cpp
// 백준 9012번
#include <iostream>
#include <vector>

using namespace std;

// solution 함수로 return값 낼 수 있도록 작성
vector<string> solution(vector<string> parenthesis) {
	vector<string> result;
	
	for (int i = 0; i < parenthesis.size(); i++) {
		// 괄호문자열의 길이가 홀수인 경우 무조건 VPS가 아님
		if (parenthesis[i].length() % 2 != 0) {
			result.push_back("NO");
			continue;
		}

		vector<char> stack;

		int j;
		for ( j = 0; j < parenthesis[i].size(); j++) {
			// '(' 이면 스택에 넣고 ')' 이면 스택의 원소를 삭제
			if (parenthesis[i][j] == '(')
				stack.push_back('(');
			else {
				if(!stack.empty())
					stack.pop_back();
				else {
					// ')'인데 스택이 비어있는 경우 짝이 맞지 않음
					break;
				}
			}
		}

		if (j == parenthesis[i].length() && stack.empty())
			result.push_back("YES");
		else	// 모든 원소를 검사했음에도 스택이 비어있지 않는 경우도 짝이 맞지 않음
			result.push_back("NO");
	}
	return result;
}

int main() {
	int T;
	cin >> T;

	vector<string> parenthesis(T);

	for (int i = 0; i < T; i++)
		cin >> parenthesis[i];

	vector<string> result = solution(parenthesis);

	for (int i = 0; i < T; i++) {
		cout << result[i] << endl;
	}
}

File: jieun^,^/18회/하_괄호.py
# 백준 9012번

def VPS(PS):  # 올바른 문자열 VPS - return 1, 아니면 - return 0
    # '('와 ')'로만 구성
    # '('가 들어오면 스택에 넣고,
    # ')'가 들어오면 스택이 isempty인지 확인하고 스택을 pop하기, 만약 empty거나, 최종 스택에 남는게 있으면 잘못됐음 -> return 0

    stack=[]
    for i in range(len(PS)):
        if PS[i] == '(':
            stack.append('(')
        
        else:   # PS[i] == ')'
            if len(stack)==0:   # 스택이 empty
                return 0        # 잘못된 문자열
            else:
                stack.pop()
        
    if len(stack)!=0: return 0  # 최종 스택에 남아있는 것이 있다면 잘못된 문자열

    return 1    # 올바른 문자열


def solution():
    T = int(input())
    PS = []
    for i in range(T):
        x=input()
        PS.append(x)

    for i in range(T):
        res = VPS(PS[i])
        if res==1:
            print("YES")
        else:
            print("NO")    


solution()